File: android/app/src/main/kotlin/com/example/flixsy/ConnectSdkPlugin.kt
package com.example.flixsy

import android.content.Context
import android.net.wifi.WifiManager
import android.util.Log
import com.connectsdk.device.ConnectableDevice
import com.connectsdk.device.ConnectableDeviceListener
import com.connectsdk.discovery.DiscoveryManager
import com.connectsdk.discovery.DiscoveryManagerListener
import com.connectsdk.service.DeviceService
import com.connectsdk.service.command.ServiceCommandError
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.Result

class ConnectSdkPlugin(private val context: Context) :
    MethodChannel.MethodCallHandler,
    EventChannel.StreamHandler,
    DiscoveryManagerListener {

    companion object {
        const val METHOD_CHANNEL = "com.flixsy.app/connect_sdk"
        const val EVENT_CHANNEL = "com.flixsy.app/connect_sdk_events"
        private const val TAG = "ConnectSdkPlugin"
    }

    private var eventSink: EventChannel.EventSink? = null
    private val discoveredDevices = mutableMapOf<String, ConnectableDevice>()
    private var connectedDevice: ConnectableDevice? = null
    private var multicastLock: WifiManager.MulticastLock? = null

    // MARK: - MethodCallHandler

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "startDiscovery" -> startDiscovery(result)
            "stopDiscovery" -> stopDiscovery(result)
            "connectToDevice" -> {
                val deviceId = call.argument<String>("deviceId")
                    ?: return result.error("INVALID_ARGS", "deviceId required", null)
                connectToDevice(deviceId, result)
            }
            "disconnect" -> disconnect(result)
            "sendKeyCommand" -> {
                val key = call.argument<String>("key")
                    ?: return result.error("INVALID_ARGS", "key required", null)
                sendKeyCommand(key, result)
            }
            "getDiscoveredDevices" -> getDiscoveredDevices(result)
            else -> result.notImplemented()
        }
    }

    // MARK: - Discovery

    private fun startDiscovery(result: Result) {
        Log.d(TAG, "startDiscovery invoked")
        acquireMulticastLock()
        DiscoveryManager.init(context.applicationContext)
        val manager = DiscoveryManager.getInstance()
        manager.pairingLevel = DiscoveryManager.PairingLevel.ON
        manager.addListener(this)
        manager.start()
        Log.d(TAG, "DiscoveryManager.start() called; listener added, multicast lock held=${multicastLock?.isHeld}")
        result.success(null)
    }

    private fun stopDiscovery(result: Result) {
        DiscoveryManager.getInstance()?.apply {
            removeListener(this@ConnectSdkPlugin)
            stop()
        }
        releaseMulticastLock()
        result.success(null)
    }

    // MARK: - Connection

    private fun connectToDevice(deviceId: String, result: Result) {
        val device = discoveredDevices[deviceId]
            ?: return result.error("CONNECTION_ERROR", "Device not found: $deviceId", null)
        device.addListener(makeDeviceListener())
        device.connect()
        connectedDevice = device
        // Connection completes asynchronously; ack the call here.
        result.success(null)
    }

    private fun disconnect(result: Result) {
        connectedDevice?.disconnect()
        connectedDevice = null
        result.success(null)
    }

    // MARK: - Commands

    private fun sendKeyCommand(key: String, result: Result) {
        val device = connectedDevice
            ?: return result.error("COMMAND_ERROR", "No device connected", null)

        val success = com.connectsdk.service.capability.listeners.ResponseListener<Any> { result.success(null) }
        val failure = com.connectsdk.service.capability.listeners.ResponseListener<ServiceCommandError> {
            result.error("COMMAND_ERROR", it?.message ?: "Key command failed", null)
        }

        // Reuse to avoid repetition in the when block
        fun keyControl() = device.getCapability(com.connectsdk.service.capability.KeyControl::class.java)
        fun volumeControl() = device.getCapability(com.connectsdk.service.capability.VolumeControl::class.java)
        fun tvControl() = device.getCapability(com.connectsdk.service.capability.TVControl::class.java)
        fun mediaControl() = device.getCapability(com.connectsdk.service.capability.MediaControl::class.java)

        when (key) {
            "up" -> keyControl()?.up(success)
            "down" -> keyControl()?.down(success)
            "left" -> keyControl()?.left(success)
            "right" -> keyControl()?.right(success)
            "ok", "select" -> keyControl()?.ok(success)
            "back" -> keyControl()?.back(success)
            "home" -> keyControl()?.home(success)
            "menu" -> tvControl()?.showMenu(success)
            "volumeUp" -> volumeControl()?.volumeUp(success)
            "volumeDown" -> volumeControl()?.volumeDown(success)
            "mute" -> volumeControl()?.setMute(true, success)
            "channelUp" -> tvControl()?.channelUp(success)
            "channelDown" -> tvControl()?.channelDown(success)
            "play" -> mediaControl()?.play(success)
            "pause" -> mediaControl()?.pause(success)
            "stop" -> mediaControl()?.stop(success)
            "rewind" -> mediaControl()?.rewind(success)
            "fastForward" -> mediaControl()?.fastForward(success)
            else -> result.error("COMMAND_ERROR", "Unknown key: $key", null)
        }
    }

    private fun getDiscoveredDevices(result: Result) {
        val devices = discoveredDevices.values.map { deviceToMap(it) }
        result.success(devices)
    }

    // MARK: - DiscoveryManagerListener

    override fun onDeviceAdded(manager: DiscoveryManager, device: ConnectableDevice) {
        Log.d(TAG, "onDeviceAdded: id=${device.id} name=${device.friendlyName} ip=${device.ipAddress} services=${device.services?.size}")
        discoveredDevices[device.id] = device
        sendEvent(mapOf("type" to "deviceFound", "device" to deviceToMap(device)))
    }

    override fun onDeviceUpdated(manager: DiscoveryManager, device: ConnectableDevice) {
        Log.d(TAG, "onDeviceUpdated: id=${device.id} name=${device.friendlyName}")
        discoveredDevices[device.id] = device
        sendEvent(mapOf("type" to "deviceUpdated", "device" to deviceToMap(device)))
    }

    override fun onDeviceRemoved(manager: DiscoveryManager, device: ConnectableDevice) {
        Log.d(TAG, "onDeviceRemoved: id=${device.id}")
        discoveredDevices.remove(device.id)
        sendEvent(mapOf("type" to "deviceLost", "deviceId" to (device.id ?: "")))
    }

    override fun onDiscoveryFailed(manager: DiscoveryManager, error: ServiceCommandError) {
        Log.e(TAG, "onDiscoveryFailed: ${error.message}")
        sendEvent(mapOf("type" to "discoveryError", "message" to (error.message ?: "Unknown error")))
    }

    // MARK: - StreamHandler

    override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
        Log.d(TAG, "EventChannel onListen — Dart attached")
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        Log.d(TAG, "EventChannel onCancel — Dart detached")
        eventSink = null
    }

    // MARK: - Device listener factory

    private fun makeDeviceListener() = object : ConnectableDeviceListener {
        override fun onDeviceReady(device: ConnectableDevice) {
            sendEvent(mapOf("type" to "deviceConnected", "device" to deviceToMap(device)))
        }

        override fun onDeviceDisconnected(device: ConnectableDevice) {
            sendEvent(mapOf(
                "type" to "deviceDisconnected",
                "deviceId" to (device.id ?: ""),
                "message" to "",
            ))
            if (connectedDevice?.id == device.id) connectedDevice = null
        }

        override fun onPairingRequired(
            device: ConnectableDevice,
            service: DeviceService,
            pairingType: DeviceService.PairingType,
        ) {
            sendEvent(mapOf("type" to "pairingRequired", "deviceId" to (device.id ?: "")))
        }

        override fun onCapabilityUpdated(
            device: ConnectableDevice,
            added: MutableList<String>,
            removed: MutableList<String>,
        ) {
            // No action needed.
        }

        override fun onConnectionFailed(device: ConnectableDevice, error: ServiceCommandError) {
            sendEvent(mapOf(
                "type" to "connectionFailed",
                "deviceId" to (device.id ?: ""),
                "message" to (error.message ?: ""),
            ))
        }
    }

    // MARK: - Helpers

    private fun deviceToMap(device: ConnectableDevice): Map<String, Any> = mapOf(
        "id" to (device.id ?: ""),
        "name" to (device.friendlyName ?: "Unknown"),
        "ipAddress" to (device.ipAddress ?: ""),
        "connected" to device.isConnected,
    )

    private fun sendEvent(payload: Map<String, Any>) {
        if (eventSink == null) {
            Log.w(TAG, "sendEvent with no eventSink — Dart isn't listening yet. payload=$payload")
        }
        eventSink?.success(payload)
    }

    // MARK: - Multicast lock (required for device discovery on Android)

    private fun acquireMulticastLock() {
        if (multicastLock == null) {
            val wifiManager = context.applicationContext
                .getSystemService(Context.WIFI_SERVICE) as WifiManager
            multicastLock = wifiManager.createMulticastLock("flixsy_connectsdk").apply {
                setReferenceCounted(true)
                acquire()
            }
        }
    }

    private fun releaseMulticastLock() {
        multicastLock?.let {
            if (it.isHeld) it.release()
        }
        multicastLock = null
    }
}

File: ios/Runner/ConnectSdkPlugin.swift
import Flutter
import UIKit
import ConnectSDK

class ConnectSdkPlugin: NSObject {

    private let methodChannel: FlutterMethodChannel
    private let eventChannel: FlutterEventChannel
    private var eventSink: FlutterEventSink?

    private var discoveredDevices: [String: ConnectableDevice] = [:]
    private var connectedDevice: ConnectableDevice?

    init(binaryMessenger: FlutterBinaryMessenger) {
        methodChannel = FlutterMethodChannel(
            name: "com.flixsy.app/connect_sdk",
            binaryMessenger: binaryMessenger
        )
        eventChannel = FlutterEventChannel(
            name: "com.flixsy.app/connect_sdk_events",
            binaryMessenger: binaryMessenger
        )
        super.init()
        methodChannel.setMethodCallHandler(handle)
        eventChannel.setStreamHandler(self)
    }

    // MARK: - Method call handler

    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "startDiscovery":
            startDiscovery(result: result)
        case "stopDiscovery":
            stopDiscovery(result: result)
        case "connectToDevice":
            guard let args = call.arguments as? [String: Any],
                  let deviceId = args["deviceId"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "deviceId required", details: nil))
                return
            }
            connectToDevice(deviceId: deviceId, result: result)
        case "disconnect":
            disconnect(result: result)
        case "sendKeyCommand":
            guard let args = call.arguments as? [String: Any],
                  let key = args["key"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "key required", details: nil))
                return
            }
            sendKeyCommand(key: key, result: result)
        case "getDiscoveredDevices":
            getDiscoveredDevices(result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Discovery

    private func startDiscovery(result: @escaping FlutterResult) {
        NSLog("[ConnectSdkPlugin] startDiscovery invoked")
        guard let manager = DiscoveryManager.shared() else {
            NSLog("[ConnectSdkPlugin] DiscoveryManager.shared() returned nil")
            result(FlutterError(code: "DISCOVERY_ERROR", message: "DiscoveryManager unavailable", details: nil))
            return
        }
        manager.delegate = self
        manager.pairingLevel = DeviceServicePairingLevelOn
        manager.startDiscovery()
        NSLog("[ConnectSdkPlugin] DiscoveryManager.startDiscovery() called; delegate=self pairingLevel=ON")

        result(nil)
    }

    private func stopDiscovery(result: @escaping FlutterResult) {
        DiscoveryManager.shared()?.stopDiscovery()
        result(nil)
    }

    // MARK: - Connection

    private func connectToDevice(deviceId: String, result: @escaping FlutterResult) {
        guard let device = discoveredDevices[deviceId] else {
            result(FlutterError(code: "CONNECTION_ERROR", message: "Device not found: \(deviceId)", details: nil))
            return
        }
        device.delegate = self
        device.connect()
        connectedDevice = device
        // Connection result comes asynchronously via ConnectableDeviceDelegate; ack here.
        result(nil)
    }

    private func disconnect(result: @escaping FlutterResult) {
        connectedDevice?.disconnect()
        connectedDevice = nil
        result(nil)
    }

    // MARK: - Commands

    private func sendKeyCommand(key: String, result: @escaping FlutterResult) {
        guard let device = connectedDevice else {
            result(FlutterError(code: "COMMAND_ERROR", message: "No device connected", details: nil))
            return
        }

        let success: SuccessBlock = { _ in DispatchQueue.main.async { result(nil) } }
        let failure: FailureBlock = { error in
            DispatchQueue.main.async {
                result(FlutterError(
                    code: "COMMAND_ERROR",
                    message: error?.localizedDescription ?? "Key command failed",
                    details: nil
                ))
            }
        }

        switch key {
        case "up":
            device.keyControl()?.up(success: success, failure: failure)
        case "down":
            device.keyControl()?.down(success: success, failure: failure)
        case "left":
            device.keyControl()?.left(success: success, failure: failure)
        case "right":
            device.keyControl()?.right(success: success, failure: failure)
        case "ok", "select":
            device.keyControl()?.ok(success: success, failure: failure)
        case "back":
            device.keyControl()?.back(success: success, failure: failure)
        case "home":
            device.keyControl()?.home(success: success, failure: failure)
        case "menu":
            result(FlutterError(code: "COMMAND_ERROR", message: "menu key not supported by ConnectSDK KeyControl", details: nil))
        case "volumeUp":
            device.volumeControl()?.volumeUp(success: success, failure: failure)
        case "volumeDown":
            device.volumeControl()?.volumeDown(success: success, failure: failure)
        case "mute":
            device.volumeControl()?.setMute(true, success: success, failure: failure)
        case "channelUp":
            device.tvControl()?.channelUp(success: success, failure: failure)
        case "channelDown":
            device.tvControl()?.channelDown(success: success, failure: failure)
        case "play":
            device.mediaControl()?.play(success: success, failure: failure)
        case "pause":
            device.mediaControl()?.pause(success: success, failure: failure)
        case "stop":
            device.mediaControl()?.stop(success: success, failure: failure)
        case "rewind":
            device.mediaControl()?.rewind(success: success, failure: failure)
        case "fastForward":
            device.mediaControl()?.fastForward(success: success, failure: failure)
        default:
            result(FlutterError(code: "COMMAND_ERROR", message: "Unknown key: \(key)", details: nil))
        }
    }

    private func getDiscoveredDevices(result: @escaping FlutterResult) {
        let devices = discoveredDevices.values.map { deviceToMap($0) }
        result(Array(devices))
    }

    // MARK: - Helpers

    private func deviceToMap(_ device: ConnectableDevice) -> [String: Any] {
        return [
            "id": device.id ?? "",
            "name": device.friendlyName ?? "Unknown",
            "ipAddress": device.address ?? "",
            "connected": device.connected,
        ]
    }

    private func sendEvent(_ payload: [String: Any]) {
        DispatchQueue.main.async {
            if self.eventSink == nil {
                NSLog("[ConnectSdkPlugin] WARNING: sendEvent with no eventSink — Dart isn't listening yet. payload=\(payload)")
            }
            self.eventSink?(payload)
        }
    }
}

// MARK: - FlutterStreamHandler

extension ConnectSdkPlugin: FlutterStreamHandler {
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        NSLog("[ConnectSdkPlugin] EventChannel onListen — Dart attached")
        eventSink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        NSLog("[ConnectSdkPlugin] EventChannel onCancel — Dart detached")
        eventSink = nil
        return nil
    }
}

// MARK: - DiscoveryManagerDelegate

extension ConnectSdkPlugin: DiscoveryManagerDelegate {
    func discoveryManager(_ manager: DiscoveryManager!, didFind device: ConnectableDevice!) {
        guard let device = device, let id = device.id else {
            NSLog("[ConnectSdkPlugin] didFind called with nil device/id")
            return
        }
        NSLog("[ConnectSdkPlugin] didFind: id=\(id) name=\(device.friendlyName ?? "?") addr=\(device.address ?? "?") services=\(device.services?.count ?? 0)")
        discoveredDevices[id] = device
        sendEvent(["type": "deviceFound", "device": deviceToMap(device)])
    }

    func discoveryManager(_ manager: DiscoveryManager!, didLose device: ConnectableDevice!) {
        guard let device = device, let id = device.id else { return }
        NSLog("[ConnectSdkPlugin] didLose: id=\(id) name=\(device.friendlyName ?? "?")")
        discoveredDevices.removeValue(forKey: id)
        sendEvent(["type": "deviceLost", "deviceId": id])
    }

    func discoveryManager(_ manager: DiscoveryManager!, didFailWithError error: Error!) {
        NSLog("[ConnectSdkPlugin] didFailWithError: \(error?.localizedDescription ?? "nil")")
        sendEvent(["type": "discoveryError", "message": error?.localizedDescription ?? "Unknown error"])
    }
}

// MARK: - ConnectableDeviceDelegate

extension ConnectSdkPlugin: ConnectableDeviceDelegate {
    func connectableDeviceReady(_ device: ConnectableDevice!) {
        guard let device = device else { return }
        sendEvent(["type": "deviceConnected", "device": deviceToMap(device)])
    }

    func connectableDeviceDisconnected(_ device: ConnectableDevice!, withError error: Error!) {
        guard let device = device else { return }
        sendEvent([
            "type": "deviceDisconnected",
            "deviceId": device.id ?? "",
            "message": error?.localizedDescription ?? "",
        ])
        if connectedDevice?.id == device.id {
            connectedDevice = nil
        }
    }

    func connectableDevice(
        _ device: ConnectableDevice!,
        service: DeviceService!,
        pairingRequiredOfType pairingType: Int32,
        withData pairingData: Any!
    ) {
        sendEvent(["type": "pairingRequired", "deviceId": device?.id ?? ""])
    }

    func connectableDevice(
        _ device: ConnectableDevice!,
        service: DeviceService!,
        pairingSucceededWithCode pairingCode: String!
    ) {
        sendEvent(["type": "pairingSucceeded", "deviceId": device?.id ?? ""])
    }

    func connectableDevice(
        _ device: ConnectableDevice!,
        service: DeviceService!,
        pairingFailedWithError error: Error!
    ) {
        sendEvent([
            "type": "pairingFailed",
            "deviceId": device?.id ?? "",
            "message": error?.localizedDescription ?? "",
        ])
    }

    func connectableDeviceConnectionSuccess(_ device: ConnectableDevice!, for service: DeviceService!) {
        // Individual service connected — no action needed; deviceReady fires when all services are ready.
    }

    func connectableDevice(_ device: ConnectableDevice!, connectionFailedWithError error: Error!) {
        sendEvent([
            "type": "connectionFailed",
            "deviceId": device?.id ?? "",
            "message": error?.localizedDescription ?? "",
        ])
    }
}
